### ISSUE TITLE
Private in public error not triggered by re-exports

### ISSUE THREAD

**Issue by Vurich** [rust-lang/rust#46681]
Minimal reproduction ([playground link](https://play.rust-lang.org/?gist=0a02f34ceb9c5d9a1934fedb1d03be53&version=stable)):

```rust
mod bad {
    pub use self::inner::test;

    mod inner {
        pub struct Foo;
        
        pub fn test() -> Foo { Foo }
    }
}

fn main() {
    // Uncommenting this results in "error[E0603]: module `inner` is private"
    // use bad::inner::Foo;

    let what = bad::test();
}
```

**Comment by petrochenkov** [comment]
This is expected behavior.
`pub use self::inner::test;` tries to reexport `fn test` which is public too (marked with `pub`), not its private parent module `inner`, so no private-in-public error happens.
On the other hand, reexporting `inner` (`pub use self::inner;`) would fail because `inner` is private (not marked with `pub`).

This situation also commonly appears in the facade pattern when a crate structures its internals using a bunch of private implementation-detail modules that contain public items that are then reexported from a single public facade module:
```rust
pub mod facade {
    pub use details1::ApiStruct;
    pub use details2::ApiTrait;
    pub use details3::api_fn;

    mod details1 {
        pub struct ApiStruct;
    }
    mod details2 {
        pub trait ApiTrait {}
    }
    mod details3 {
        pub fn api_fn() {}
    }
}
```
